dbms/src/Interpreters/Join.cpp

```diff
@@ -1098,15 +1152,32 @@ void NO_INLINE joinBlockImplTypeCase(
     IColumn::Offset & current_offset,
     std::unique_ptr<IColumn::Offsets> & offsets_to_replicate,
     const std::vector<size_t> & right_indexes,
-    const TiDB::TiDBCollators & collators)
+    const TiDB::TiDBCollators & collators,
+    bool enable_fine_grained_shuffle,
+    size_t fine_grained_shuffle_count)
 {
     size_t num_columns_to_add = right_indexes.size();
 
     KeyGetter key_getter(key_columns, key_sizes, collators);
     std::vector<std::string> sort_key_containers;
     sort_key_containers.resize(key_columns.size());
     Arena pool;
+    WeakHash32 shuffle_hash(0); /// reproduce hash values in FinedGrainedShuffleWriter
+    if (enable_fine_grained_shuffle && rows > 0)
+    {
+        /// TODO: consider adding a virtual column in Sender side to avoid computing cost and potential inconsistency by heterogeneous envs
+        /// Note: 1. Not sure, if inconsistency will do happen in heterogeneous envs
```

Comment:
Why there maybe problems in heterogeneous envs?

Reply:
Add comments to make it clear.